src/localplayer.cpp

```diff
@@ -388,31 +301,79 @@ void LocalPlayer::move(f32 dtime, Environment *env, f32 pos_max_d,
 				continue;
 
 			min_distance_f = distance_f;
-			new_sneak_node = p;
+			m_sneak_node = p;
+			m_sneak_node_exists = true;
 		}
 
-		bool sneak_node_found = (min_distance_f < 100000.0 * BS);
-		m_sneak_node = new_sneak_node;
-		m_sneak_node_exists = sneak_node_found;
-
-		if (sneak_node_found) {
+		if (m_sneak_node_exists) {
 			// Update saved top bounding box of sneak node
-			MapNode n = map->getNodeNoEx(m_sneak_node);
+			node = map->getNodeNoEx(m_sneak_node);
 			std::vector<aabb3f> nodeboxes;
-			n.getCollisionBoxes(nodemgr, &nodeboxes);
+			node.getCollisionBoxes(nodemgr, &nodeboxes);
 			m_sneak_node_bb_top = getTopBoundingBox(nodeboxes);
 
-			m_sneak_ladder_detected = physics_override_sneak_glitch &&
-					detectSneakLadder(map, nodemgr, floatToInt(position, BS));
-		} else {
-			m_sneak_ladder_detected = false;
+			if (physics_override_sneak_glitch) {
+				// Detect sneak ladder:
+				// Node two meters above sneak node must be solid
+				node = map->getNodeNoEx(m_sneak_node + v3s16(0,2,0),
+					&is_valid_position);
+				if (is_valid_position && nodemgr->get(node).walkable) {
+					// Node three meters above: must be non-solid
+					node = map->getNodeNoEx(m_sneak_node + v3s16(0,3,0),
+						&is_valid_position);
+					m_sneak_ladder_detected = is_valid_position &&
+						!nodemgr->get(node).walkable;
+				}
+			}
 		}
 	}
 
+	if (m_sneak_ladder_detected) {
+		// restore legacy behaviour (this makes the m_speed.Y hack necessary)
+		sneak_max = v3f(0.4 * BS, 0, 0.4 * BS);
+	}
```

Comment:
It's very confusing to move this code far away from the definition of `sneak_max`